feat: query cache provider

jackmellis opened the pull request:
the query cache now comes from a context, meaning users can easily override and control the cache being used inside their application

see https://github.com/tannerlinsley/react-query/issues/460

this is a basic use case of creating an isolated cache for testing

in component:
```jsx
const MyComponent = () => {
  const queryCache = useQueryCache();

  const [ mutate ] = useMutation(someFn, {
    onSettled() {
      queryCache.refetchQueries('basket');
    }
  });
  return ( ... );
};
```

in unit test:
```jsx
import { makeQueryCache, QueryCacheProvider } from 'react-query';

const cache = makeQueryCache();
sinon.spy(cache, 'refetchQueries');


mount(
  <QueryCacheProvider queryCache={cache}>
    <MyComponent/>
  </QueryCacheProvider>
);

// ...

await invokeSomeMutation(); // after mutating, we should refetch the basket

expect(cache.refetchQueries).to.be.calledWith('basket');
```

Commit:
feat: query cache provider

the query cache now comes from a context, meaning users can easily override and control the cache being used inside their application

tannerlinsley commented:
Nice! I would love to see the tests converted to use this. Thoughts?

jackmellis commented:
Sure I can update them if you like

tannerlinsley commented:
If it’s not too much trouble! 👍

jackmellis commented:
I've updated the tests. I sometimes get a warning about overlapping `act()` calls. I can't figure out why and I'm not sure if it was happening before or if I've uncovered something...

tannerlinsley commented on src/setFocusHandler.js:
```diff
@@ -9,7 +9,7 @@ const onWindowFocus = () => {
 
   if (isDocumentVisible() && isOnline()) {
-    queryCache
+    defaultQueryCacheRef.current
```
Hmmm. Does this feel funny to you? What if a user had multiple query caches in their app? How would they hook up to this? Should multiple caches be allowed?

jackmellis replied:
yes it definitely felt funny! But as you can set up the focus handler outside of any sort of react context, it's very hard to determine which cache to use. I used the same pattern as the config where there is a `defaultConfigRef` that the provider writes over. the focus handler actually uses the config ref as well so it's already a bit of a grey area...

jackmellis replied:
for reference https://github.com/tannerlinsley/react-query/blob/master/src/setFocusHandler.js#L9

Definitely open to cleaner solutions!

Ephem replied:
Since this specific config value is "global", it kind of do make sense to use the latest one. For caches, maybe we should keep track of _all active_ caches and update them all? I think that should make it possible to get rid of `defaultQueryCacheRef`?

jackmellis replied:
I've removed the `defaultQueryCacheRef` in favour of a list of active caches

Ephem commented:
Nice work! This also relates to server rendering as per #461, if you want I can take a closer look and test it from that perspective tomorrow. 😄

tannerlinsley commented:
@Ephem Awesome, since this is clearly cross-cutting, I'll let you test before proceeding.

Also, just to confirm, this will all still be backwards compatible right? With the main goals being:
- Truly isolated tests
- SSR isolation

Confirm?

Ephem commented on src/queryCache.js:
```diff
@@ -14,6 +15,25 @@ import { defaultConfigRef } from './config'
+  return React.useContext(queryCacheContext)
+}
+
+export function ReactQueryCacheProvider({ queryCache, children }) {
```
What happens if `queryCache` changes here?

* If the old prop value was a `queryCache`, it's probably up to the parent to clean it up if necessary, it would be unsafe for this Provider to do so since it might still be in use
* If the old value was undefined, the old cache was created in this Provider, so this provider needs to clean it up

This relates to the more general issue of cleanup which I'll add a separate comment for.

jackmellis replied:
yup I've addressed this now (per my other comments)

Ephem reviewed (changes requested):
First of all, very nice work on this, I like the approach you are taking!

Just so my view on this PR is clear I think the goal should be to support explicit caches for testing, using different caches for different parts of the application/separate React roots and laying the foundation to build improved SSR support on top, but not change SSR behaviour in this PR.

**Tests**

* All internal tests are now updated to the new approach, which is great, but this also means the old approach is currently untested. Maybe add a few key tests targeting using the global cache, without a Provider?
* Maybe add a few tests using multiple caches/providers?

**Query cleanup**

I think an existing major issue right now is cache cleanup. I think the issues #432 and #460 are both related to this and also the act warnings in this PR. I think this could and should be fixed in this PR since the problem might actually become worse otherwise.

EDIT: This could just as well be done in a separate PR before this is merged.

Let me expand:

I'm pretty sure these problems stem from query timeouts not being cleared.

```js
cache.clear = () => {
  cache.queries = {}
  notifyGlobalListeners()
}
```

The old cache clear function does not in any way remove the pending query timeouts for staleness etc. To get rid of these problems, I think calling `cache.clear()` should also synchronously loop through all queries and remove _all_ timeouts related to that query.

* If a user creates a cache manually with `makeQueryCache`, they own the cache and are responsible for calling `cache.clear()`.
* If `react-query` creates a cache, the library owns that cache and needs to clean it up if it is no longer used (for example, in `ReactQueryCacheProvider`).

**Serverside rendering**

This approach looks great from a SSR-perspective and this PR implements a key part of what is needed to improve that story. There is more work that is needed on top, but that should be done in a follow up and I see nothing that wouldn't be compatible!

tannerlinsley commented on src/queryCache.js:
```diff
@@ -14,6 +15,25 @@ import { defaultConfigRef } from './config'
+}
+
+export function ReactQueryCacheProvider({ queryCache, children }) {
+  defaultQueryCacheRef.current = queryCache || makeQueryCache()
```
It should only create it once. That looks like it will be created on every render (if the queryCache isn't passed)

jackmellis replied:
I've cleaned this logic up now so it tracks queryCache instances between renders

Commit:
fix: track multiple query caches

this commit does the follwing:
track all active query caches added via the provider, handles mounting/unmounting
if creating a cache, ensure we only do it once
when unmounting an internally-created cache, make sure to clear it first
setFocusHandler now loops through all of the currently-active caches

Commit:
test: add in-depth tests around query cache providers

Commit:
style: format touched files

jackmellis commented:
> First of all, very nice work on this, I like the approach you are taking!
> 
> Just so my view on this PR is clear I think the goal should be to support explicit caches for testing, using different caches for different parts of the application/separate React roots and laying the foundation to build improved SSR support on top, but not change SSR behaviour in this PR.
> 
> **Tests**
> 
> * All internal tests are now updated to the new approach, which is great, but this also means the old approach is currently untested. Maybe add a few key tests targeting using the global cache, without a Provider?
> * Maybe add a few tests using multiple caches/providers?
> 
> **Query cleanup**
> 
> I think an existing major issue right now is cache cleanup. I think the issues #432 and #460 are both related to this and also the act warnings in this PR. I think this could and should be fixed in this PR since the problem might actually become worse otherwise.
> 
> EDIT: This could just as well be done in a separate PR before this is merged.
> 
> Let me expand:
> 
> I'm pretty sure these problems stem from query timeouts not being cleared.
> 
> ```js
> cache.clear = () => {
>   cache.queries = {}
>   notifyGlobalListeners()
> }
> ```
> 
> The old cache clear function does not in any way remove the pending query timeouts for staleness etc. To get rid of these problems, I think calling `cache.clear()` should also synchronously loop through all queries and remove _all_ timeouts related to that query.
> 
> * If a user creates a cache manually with `makeQueryCache`, they own the cache and are responsible for calling `cache.clear()`.
> * If `react-query` creates a cache, the library owns that cache and needs to clean it up if it is no longer used (for example, in `ReactQueryCacheProvider`).
> 
> **Serverside rendering**
> 
> This approach looks great from a SSR-perspective and this PR implements a key part of what is needed to improve that story. There is more work that is needed on top, but that should be done in a follow up and I see nothing that wouldn't be compatible!

I've added tests to show the global/default cache still works and a few tests for different caching scenarios

The provider now tracks its queryCache value and if it's a self-created one, it cleans it up and removes it on unmount. If the cache is manually created, it's up to the creator to manage cleanup.

I haven't touched on the in-depth cache cleanup (cancelling timeouts etc.) as I don't want to lose the focus of this PR. Happy to take this on separately though if nobody is working on it...

Commit:
fixup! style: format touched files

Ephem commented:
Changes and tests looks great to me, really great work!

I totally agree with addressing cache cleanup in a separate PR. Since it's possible that using a lot of Providers in tests could worsen the existing problems (more timeouts) I think it should probably be addressed before a release.

tannerlinsley commented on README.md:
```diff
@@ -2582,6 +2596,29 @@ function App() {
 
+## `ReactQueryCacheProvider`
+
+`ReactQueryCacheProvider` is an optional provider component for explicitly setting the query cache used by `useQuery`. This is useful for creating component-level caches that are not completely global, as well as making truly isolated unit tests.
```
```suggestion
`ReactQueryCacheProvider` is an optional provider component for explicitly setting the query cache used by React Query. This is useful for creating component-level caches that are not completely global, as well as making truly isolated unit tests.
```

jackmellis committed:
docs: correct readme description

Co-authored-by: Tanner Linsley <tannerlinsley@gmail.com>

tannerlinsley commented:
:tada: This PR is included in version 1.5.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/react-query/v/1.5.0)
- [GitHub release](https://github.com/tannerlinsley/react-query/releases/tag/v1.5.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
